tools/lsp/server_loop.rs

```diff
@@ -884,6 +892,93 @@ pub async fn reload_document(
     Ok(())
 }
 
+fn extract_rust_macro(content: String) -> Option<String> {
+    let mut begin = 0;
+    loop {
+        if let Some(m) = content[begin..].find("slint") {
+            begin += m + 5;
+            while content[begin..].starts_with(' ') {
+                begin += 1;
+            }
+            if !content[begin..].starts_with('!') {
+                continue;
+            }
+            begin += 1;
+            while content[begin..].starts_with(' ') {
+                begin += 1;
+            }
+            if !content[begin..].starts_with('{') {
+                continue;
+            }
+            begin += 1;
+            break;
+        } else {
+            // No macro found, just return
+            return None;
+        }
+    }
+
+    // Now find the matching '}'
+    // Technically, we should be lexing rust, not slint
+    let mut state = i_slint_compiler::lexer::LexState::default();
+    let mut end = begin;
+    let mut level = 0;
+    while !content[end..].is_empty() {
+        let len = match i_slint_compiler::parser::lex_next_token(&content[end..], &mut state) {
+            Some((len, i_slint_compiler::parser::SyntaxKind::LBrace)) => {
+                level += 1;
+                len
+            }
+            Some((_, i_slint_compiler::parser::SyntaxKind::RBrace)) if level == 0 => {
+                break;
+            }
+            Some((len, i_slint_compiler::parser::SyntaxKind::RBrace)) => {
+                level -= 1;
+                len
+            }
+            Some((len, _)) => len,
+            None => {
+                // Lex error
+                break;
+            }
+        };
+        if len == 0 {
+            break; // Shouldn't happen
+        }
+        end += len;
+    }
+    let mut bytes = content.into_bytes();
+    for c in &mut bytes[..begin] {
+        if *c != b'\n' {
+            *c = b' '
+        }
+    }
+    for c in &mut bytes[end..] {
+        if *c != b'\n' {
+            *c = b' '
+        }
+    }
+    Some(String::from_utf8(bytes).expect("We just added spaces"))
+}
+
+#[test]
+fn test_extract_rust_macro() {
+    assert_eq!(extract_rust_macro("\nslint{!{}}".into()), None);
```

Comment:
Might be worth to add a test for the scenario of an open macro, since that'll happen every time the user starts (unless completed).